src/main/java/com/vmware/vipclient/i18n/base/cache/MessageCache.java

```diff
@@ -4,18 +4,27 @@
  */
 package com.vmware.vipclient.i18n.base.cache;
 
+import java.util.HashMap;
 import java.util.LinkedHashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.Set;
 
+import com.vmware.vipclient.i18n.messages.api.url.URLUtils;
+
 public class MessageCache implements Cache {
     private String                           id                  = "cache-default";
 
     private long                             expiredTime         = 864000000;                                       // 240hr
     private long                             lastClean           = System.currentTimeMillis();
 
-    private Map<String, Map<String, String>> cachedComponentsMap = new LinkedHashMap<String, Map<String, String>>();
-
+    private final Map<String, Map<String, String>> cachedComponentsMap = new LinkedHashMap<String, Map<String, String>>();
+    private final Map<String, Map<String, Object>> cacheProperties = new LinkedHashMap<String, Map<String, Object>>();
```

Comment:
I think it's better to have another container ( or a class similar to MessageCache) to maintain the data of 'cacheProperties', it can find etag and expiration from the container by cachekey.

Basically, MessageCache is for maintaining cached data pool and cacheProperties is for maintaining the etag-expiration. For any searching data from MessageCache, it will firstly detect if expired from  cacheProperties: if yes it should get the etag and make request to remote service, if not it should return cached data from MessageCache.

Reply:
I will wait for finalizing this to provide more comments.